source/components/ImportControls.js

```diff
@@ -0,0 +1,116 @@
+import React from 'react'
+
+const CUSTOM_LABEL = 'Upload custom'
+
+export default class ImportControls extends React.Component {
+  constructor(props) {
+    super(props)
+
+    this.state = {
+      selectedIndex: 0,
+      usingUpload: false,
+      uploadedFilename: '',
+    }
+
+    this._onFileUpload = this._onFileUpload.bind(this)
+    this._resetUpload = this._resetUpload.bind(this)
+  }
+
+  _onFileUpload(event) {
+    const file = event.target.files[0]
+    const reader = new FileReader()
+    reader.onload = readEvent => {
+      const topoJson = JSON.parse(readEvent.target.result)
```

Comment:
I just started doing that and realized that we don't have a good way to communicate errors in custom data—ditto for CSV. I think it may be worth revisiting…